node/cli/src/command.rs

```diff
@@ -0,0 +1,797 @@
+// Copyright 2019-2021 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! This module constructs and executes the appropriate service components for the given subcommand
+
+use crate::cli::{Cli, RelayChainCli, Subcommand};
+use cumulus_client_service::genesis::generate_genesis_block;
+use cumulus_primitives_core::ParaId;
+use log::info;
+use parity_scale_codec::Encode;
+use polkadot_parachain::primitives::AccountIdConversion;
+use polkadot_service::RococoChainSpec;
+use sc_cli::{
+	ChainSpec, CliConfiguration, DefaultConfigurationValues, ImportParams, KeystoreParams,
+	NetworkParams, Result, RuntimeVersion, SharedParams, SubstrateCli,
+};
+use sc_service::config::{BasePath, PrometheusConfig};
+use service::{chain_spec, IdentifyVariant};
+use sp_core::hexdisplay::HexDisplay;
+use sp_core::H160;
+use sp_runtime::traits::Block as _;
+#[cfg(feature = "with-moonbase-runtime")]
+use std::str::FromStr;
+use std::{io::Write, net::SocketAddr};
+
+#[cfg(not(feature = "with-moonriver-runtime"))]
+use cli_opt::RpcParams;
+
+#[cfg(not(feature = "with-moonbase-runtime"))]
+use cli_opt::MISSING_MOONBASE;
+#[cfg(not(feature = "with-moonbeam-runtime"))]
+use cli_opt::MISSING_MOONBEAM;
+#[cfg(not(feature = "with-moonriver-runtime"))]
+use cli_opt::MISSING_MOONRIVER;
+
+fn load_spec(
+	id: &str,
+	// `para_id` only used on with-moonbase-runtime feature.
+	#[allow(unused_variables)] para_id: ParaId,
+) -> std::result::Result<Box<dyn sc_service::ChainSpec>, String> {
+	if id.is_empty() {
+		return Err("Not specific which chain to run.".into());
+	}
+	println!("load_spec {:?}", id);
+	Ok(match id {
```

Comment:
We'll have to consider how to handle `--chain custom-chain.json` where it isn't clear what runtime is supposed to be used. Polkadot bases it off the beginning of the file name. It also has CLI flags for this, but I've never gotten them to work.

Reply:
load_spec is WIP